### ISSUE TITLE
`node -d` transfers hidden nodes even though they are in the destination desktop.

### ISSUE THREAD

**Issue by emanuele6** [baskerville/bspwm#1310]
Running `node -d` on an hidden node will transfer that node to the destination desktop even though that node is already in the destination desktop (as if the source desktop was another desktop).

How to reproduce:
test1:
* open 2 tiled terminal windows;
* run `bspc node -g hidden=on -g hidden=off -f` in the first terminal;
* RESULT: the node will disappear and reappear in the same position. (ok)

test2:
* open 2 tiled terminal windows;
* run `bspc node -d focused; echo "$?"` in the first terminal;
* RESULT: `1`, the command failed because the destination desktop and the source desktop are the same desktop. (ok)

test3:
* open 2 tiled terminal windows;
* run `bspc node -g hidden=on -d focused -g hidden=off -f` in the first terminal;
* RESULT: the node will disappear and reappear as the second child of the root node; `node -d` succeeded and the node's position in the tree was changed. (ko)
* EXPECTED RESULT: the command chain should have stopped at `-g hidden=on` (because `-d focused` should have failed) or, at least, `-d focused` should have been a no-op and shouldn't have rearranged the node (i.e. same result as test1).  

Temporary solution: split the chain and use `bspc node "$node.!local" -d focused` instead of `bspc node "$node" -d focused`.

**Comment by ortango** [comment]
do you think it is enough to just add a check for `ds == dd` [@ that first check in transfer_node()](https://github.com/baskerville/bspwm/blob/e22d0fad23e0e85b401be69f2360a1c3a0767921/src/tree.c#L1610)?

that is a nice find.

**Comment by emanuele6** [comment]
> do you think it is enough to just add a check for `ds == dd` [@ that first check in transfer_node()](https://github.com/baskerville/bspwm/blob/e22d0fad23e0e85b401be69f2360a1c3a0767921/src/tree.c#L1610)?

I'm not sure since there is this block of code:

https://github.com/baskerville/bspwm/blob/e22d0fad23e0e85b401be69f2360a1c3a0767921/src/tree.c#L1641-L1650

That block was introduced with bfded1effccbed8f76fe5fbbe11c3aca665fd9c4.

There actually used to be a `dd == ds` check in `transfer_node()`, but it was removed with that same commit.
https://github.com/baskerville/bspwm/blob/9729efc422ff5f95dda0ec7cdf44115970fe2ee0/tree.c#L822-L825

The code in `transfer_node()` looked complex to me and I have not had time to study it yet so I can't tell if adding that check (and removing that block of code) would be a good solution.

**Comment by emanuele6** [comment]
I just noticed that `bspc node "$node" -d focused` has the a similar behaviour to `bspc node "$node" -n focused`.

I think this is intended.

`bspc node -d focused` doesn't fail because the destination desktop is the same as the source desktop, it fails because `bspc node -n focused` would fail too.

test4:

* open 2 tiled terminal windows;
* run `bspc node '@brother' -d focused ; echo "$?"` in the first terminal.
* the command had an effect and `0` gets printed to stdout.

This behaviour is intended; I was just misunderstanding when `node -d` should fail. I'm closing the issue.

**Comment by ortango** [comment]
are you sure that it is intended? because i wrote this script yesterday for testing and it seems to behave weirdly:

    #test-d.sh
    tn=$(bspc query -N -n)
    td=$(bspc query -D -n)
    
    if ! bspc node $1 -f; then
        echo "couldn't focus $1" >&2
        exit 1
    fi
    bspc node "$tn" -d "$td"
    echo $?

i had three windows open, two on the focused desktop, and one on an active desktop. if i run `test-d.sh next.active.window.\!hidden` the transfer will fail like it should; same thing if you tried `@parent` or `focused`. if i run `test-d.sh next.local.window.\!hidden` instead - the transfer happens (i don't think it should).

**Comment by emanuele6** [comment]
>  i'm not sure that i like the behavior though, because now i feel like i need to check which desktop a node is on before i use --to-desktop (so i don't accidentally break the tree at least)

Yeah, I feel the same way about `node -n`'s and `node -m`'s behaviour. I thought they were supposed to fail if the target node is in the destination desktop/monitor (or, at least, do nothing). I only noticed this yesterday because I noticed that my hide/unhide script that I wrote a long time ago and have used for a long time didn't work as I expected when hiding tiled windows (I usually use it only on floating windows so I never noticed it modified the tree).

Anyhow I don't know if this behaviour is worth changing since it has been this way for a long time, it allows `node -m`, `node -d`, and `node -n` to share code and it doesn't seem to have bugs.

Then again, It might be worth changing since I think I will have to perform some checks everytime I use `node -d` or `node -m` on a non-focused node now. (but I don't do this often anyway to be fair.)
